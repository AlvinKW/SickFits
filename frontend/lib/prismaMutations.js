import gql from 'graphql-tag';

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION (
		$title: String!
		$description: String!
		$image: String
		$largeImage: String
		$price: Int!
	) {
		createItem(
			title: $title
			description: $description
			image: $image
			largeImage: $largeImage
			price: $price
		) {
			id
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		createOrder(token: $token) {
			id
			total
			charge
			items {
				id
				title
			}
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const RESET_PASSWORD_MUTATION = gql`
	mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
		resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
			id
			name
			email
		}
	}
`;

export const SIGN_IN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signIn(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGN_OUT_MUTATION = gql`
	mutation SIGN_OUT_MUTATION {
		signOut {
			message
		}
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
		signUp(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const TOGGLE_CART_MUTATION = gql`
	mutation TOGGLE_CART_MUTATION {
		toggleCart @client
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation UPDATE_ITEM_MUTATION (
		$id: ID!
		$title: String
		$description: String
		$price: Int
	) {
		updateItem(
			id: $id
			title: $title
			description: $description
			price: $price
		) {
			id
			title
			description
			price
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userID: ID!) {
		updatePermissions(permissions: $permissions, userID: $userID) {
			id
			name
			email
			permissions
		}
	}
`;
