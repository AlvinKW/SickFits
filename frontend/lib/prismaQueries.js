import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
	query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
		items(orderBy: createdAt_DESC, skip: $skip, first: $first) {
			id
			title
			description
			image
			largeImage
			price
		}
	}
`;

export const ALL_USERS_QUERY = gql`
	query ALL_USERS_QUERY {
		users {
			id
			name
			email
			permissions
		}
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CURRENT_USER_QUERY {
		me {
			id
			name
			email
			permissions
			orders {
				id
			}
			cart {
				id
				quantity
				item {
					id
					title
					description
					image
					price
				}
			}
		}
	}
`;

export const LOCAL_STATE_QUERY = gql`
	query LOCAL_STATE_QUERY {
		cartOpen @client
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

export const SEARCH_ITEMS_QUERY = gql`
	query SEARCH_ITEMS_QUERY($searchTerm: String!) {
		items(where: { OR: [{ title_contains: $searchTerm }, { description_contains: $searchTerm }] }) {
			id
			title
			image
		}
	}
`;

export const SINGLE_ITEM_QUERY = gql`
	query SINGLE_ITEM_QUERY($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			description
			largeImage
			price
		}
	}
`;

export const SINGLE_ORDER_QUERY = gql`
	query SINGLE_ORDER_QUERY($id: ID!) {
		order(id: $id) {
			id
			total
			charge
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
			user {
				id
			}
		}
	}
`;

export const USER_ORDERS_QUERY = gql`
	query USER_ORDERS_QUERY {
		orders(orderBy: createdAt_DESC) {
			id
			total
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
		}
	}
`;
