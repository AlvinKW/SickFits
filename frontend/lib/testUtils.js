import casual from 'casual';

casual.seed(777);

const mockItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'Dogs are the best!',
	description: 'Dogs',
	largeImage: 'dog.jpg',
});

const mockUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const mockOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const mockOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [mockOrderItem(), mockOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: mockUser(),
});

const mockCartItem = overrides => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: mockItem(),
	user: mockUser(),
	...overrides,
});

class MockLocalStorage {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

export {
	MockLocalStorage,
	mockItem,
	mockUser,
	mockCartItem,
	mockOrder,
	mockOrderItem,
};
