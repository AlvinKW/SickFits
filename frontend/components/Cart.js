import React from 'react';
import styled from 'styled-components';
import { adopt } from 'react-adopt';
import { Query, Mutation } from 'react-apollo';

import { LOCAL_STATE_QUERY } from '../lib/prismaQueries';
import { TOGGLE_CART_MUTATION } from '../lib/prismaMutations';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';

import User from './User';
import CartItem from './CartItem';
import TakeMyMoney from './TakeMyMoney';
import SickButton from './SickButton';

const StyledCart = styled.div`
	position: fixed;
	display: grid;
	grid-template-rows: auto 1fr auto;
	height: 100%;
	width: 40%;
	min-width: 500px;
	padding: 20px;
	top: 0;
	bottom: 0;
	right: 0;
	box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.2);
	background: white;
	transform: translateX(100%);
	transition: all 0.3s;
	z-index: 5;
	${props => props.open && 'transform: translateX(0);'};

	header {
		margin-bottom: 2rem;
		padding-bottom: 2rem;
		border-bottom: 5px solid ${props => props.theme.black};
	}

	footer {
		display: grid;
		grid-template-columns: auto auto;
		align-items: center;
		margin-top: 2rem;
		padding-top: 2rem;
		border-top: 10px double ${props => props.theme.black};
		font-size: 3rem;
		font-weight: 900;

		p {
			margin: 0;
		}
	}

	ul {
		margin: 0;
		padding: 0;
		list-style: none;
		overflow: scroll;
	}

	@media only screen and (max-width: 600px) {
		min-width: 325px;
	}
`;

const CloseButton = styled.button`
	position: absolute;
	right: 0;
	border: 0;
	font-size: 3rem;
	background: black;
	color: white;
	z-index: 2;
	cursor: pointer;
`;

const Supreme = styled.h3`
	display: inline-block;
	margin: 0;
	padding: 4px 5px;
	font-size: 4rem;
	background: ${props => props.theme.red};
	color: white;
	transform: skew(-3deg);
`;

/* eslint-disable */
const Composed = adopt({
	user: ({ render }) => <User>{render}</User>,
	localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
	toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
});
/* eslint-enable */

function Cart() {
	return (
		<Composed>
			{({ user, localState, toggleCart }) => {
				const me = user.data.me;
				if (!me) { return null; }

				return (
					<StyledCart open={localState.data.cartOpen}>
						<header>
							<CloseButton onClick={toggleCart} title="close">&times;</CloseButton>
							<Supreme>Your Cart</Supreme>
							<p>You Have {me.cart.length} Item{me.cart.length === 1 ? '' : 's'} In Your Cart.</p>
						</header>
						<ul>
							{me.cart.map(cartItem => <CartItem key={cartItem.id} cartItem={cartItem} />)}
						</ul>
						<footer>
							<p>{formatMoney(calcTotalPrice(me.cart))}</p>
							{me.cart.length && (
								<TakeMyMoney>
									<SickButton>Checkout</SickButton>
								</TakeMyMoney>
							)}
						</footer>
					</StyledCart>
				);
			}}
		</Composed>
	);
}

export default Cart;
