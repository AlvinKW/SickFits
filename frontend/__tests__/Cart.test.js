import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';

import { mockUser, mockCartItem } from '../lib/testUtils';
import { CURRENT_USER_QUERY, LOCAL_STATE_QUERY } from '../lib/prismaQueries';
import Cart from '../components/Cart';

const mocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: {
				me: {
					...mockUser(),
					cart: [mockCartItem()],
				},
			},
		},
	},
	{
		request: {
			query: LOCAL_STATE_QUERY,
		},
		result: {
			data: {
				cartOpen: true,
			},
		},
	},
];

describe('<Cart />', () => {
	it('renders and matches the snapshot', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<Cart />
			</MockedProvider>
		);

		await wait();
		wrapper.update();
		const header = wrapper.find('header');
		expect(toJSON(header)).toMatchSnapshot();
		expect(wrapper.find('CartItem')).toHaveLength(1);
	});
});
