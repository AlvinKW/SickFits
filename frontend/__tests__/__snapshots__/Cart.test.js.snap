// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders and matches the snapshot 1`] = `
<header>
  <Cart__CloseButton
    onClick={[Function]}
    title="close"
  >
    <button
      className="Cart__CloseButton-sc-19s6z00-1 fCUFSU"
      onClick={[Function]}
      title="close"
    >
      Ã—
    </button>
  </Cart__CloseButton>
  <Cart__Supreme>
    <h3
      className="Cart__Supreme-sc-19s6z00-2 caCbmK"
    >
      Your Cart
    </h3>
  </Cart__Supreme>
  <p>
    You Have 
    1
     Item
     In Your Cart.
  </p>
</header>
`;
